pkg/boot/gsurv/gsurv.go

```diff
@@ -0,0 +1,372 @@
+		panic(err)
+	}
+	response, err := root.NewResponse()
+	if err != nil {
+		return nil, err
+	}
+
+	response.SetNamespace(ns)
+
+	if cab, ok := peerstore.GetCertifiedAddrBook(gsurv.h.Peerstore()); ok {
+		env := cab.GetPeerRecord(gsurv.h.ID())
+		rec, err := env.Marshal()
+		if err != nil {
+			return nil, err
+		}
+		response.SetEnvelope(rec)
+	}
+	return root.Message().MarshalPacked()
+}
+
+func (gsurv *GSurv) handleMudpResponse(response cpGSurv.GSurvResponse) {
+	var (
+		finder chan peer.AddrInfo
+		rec    peer.PeerRecord
+		err    error
+		ok     bool
+	)
+
+	gsurv.mu.Lock()
+	defer gsurv.mu.Unlock()
+
+	ns, err := response.Namespace()
+	if err != nil {
+		return
+	}
+
+	if finder, ok = gsurv.mustFind[ns]; !ok {
+		return
+	}
+
+	envelope, err := response.Envelope()
+	if err != nil {
+		return
+	}
+
+	if _, err = record.ConsumeTypedEnvelope(envelope, &rec); err != nil {
+		return
+	}
+	select {
+	case finder <- peer.AddrInfo{ID: rec.PeerID, Addrs: rec.Addrs}:
+	default:
+		return
+	}
+}
+
+func (gsurv *GSurv) Advertise(ctx context.Context, ns string, opt ...discovery.Option) (time.Duration, error) {
+	if err := gsurv.err.Load(); err != nil {
+		return 0, err.(error)
+	}
+
+	gsurv.mu.Lock()
+	defer gsurv.mu.Unlock()
+
+	opts, err := gsurv.options(ns, opt)
+	if err != nil {
+		return 0, err
+	}
+
+	if ttlChan, ok := gsurv.mustAdvertise[ns]; ok {
+		ttlChan <- opts.Ttl
+	} else {
+		resetTtl := make(chan time.Duration)
+		gsurv.mustAdvertise[ns] = resetTtl
+		go gsurv.trackAdvertise(ns, resetTtl, opts.Ttl)
+	}
+	return opts.Ttl, nil
+}
+
+func (gsurv *GSurv) options(ns string, opt []discovery.Option) (opts *discovery.Options, err error) {
+	opts = &discovery.Options{}
+	if err = opts.Apply(opt...); err == nil && opts.Ttl == 0 {
+		opts.Ttl = discTTL
+	}
+
+	return
+}
+
+func (gsurv *GSurv) trackAdvertise(ns string, resetTtl chan time.Duration, ttl time.Duration) {
+	timer := time.NewTimer(ttl)
+	defer timer.Stop()
+	for {
+		select {
+		case <-timer.C:
+			gsurv.mu.Lock()
+
+			select { // check again TTL after acquiring lock
+			case ttl := <-resetTtl:
+				timer.Reset(ttl)
+				gsurv.mu.Unlock()
+			default:
+				close(resetTtl)
+				delete(gsurv.mustAdvertise, ns)
+				gsurv.mu.Unlock()
+				return
+			}
+		case ttl := <-resetTtl:
+			timer.Reset(ttl)
+		}
+	}
+}
+
+func (gsurv *GSurv) FindPeers(ctx context.Context, ns string, opt ...discovery.Option) (<-chan peer.AddrInfo, error) {
+	if err := gsurv.err.Load(); err != nil {
+		return nil, err.(error)
+	}
+
+	gsurv.mu.Lock()
+	defer gsurv.mu.Unlock()
+
+	var (
+		opts *discovery.Options
+		dist uint8
+		err  error
+	)
+
+	opts, err = gsurv.options(ns, opt)
+	if err != nil {
+		return nil, err
+	}
+
+	if b, ok := opts.Other["distance"]; ok {
+		dist, err = conv.Uint8(b)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		dist = discDist
+	}
+
+	request, err := gsurv.buildRequest(ns, dist)
+	if err != nil {
+		return nil, err
+	}
+
+	finder := make(chan peer.AddrInfo, opts.Limit)
```

Comment:
I don't think the channel capacity is an adequate way of limiting responses, since a consumer is popping items off the queue concurrently.  Here's the pattern I usually use:

```go
out := make(chan peer.AddrInfo, 1)
go func() {
    defer close(out)

    for {
        select {
        case info := <-out:
            select {
            case out <- info:
                if opts.Limit--; opts.Limit == 0 {  // handles default value gracefully
                    return
                }

            case <-ctx.Done():
            }

        case <-ctx.Done():  // this is the contexted passed into FindPeers
            return
        }
    }
}()

return out, nil
```